flink-runtime/src/test/java/org/apache/flink/runtime/util/profiler/ProfilingServiceTest.java

```diff
@@ -0,0 +1,170 @@
+    }
+
+    @BeforeEach
+    void setUp(@TempDir Path tempDir) {
+        configs.set(RestOptions.PROFILING_RESULT_DIR, tempDir.toString());
+        profilingService = ProfilingService.getInstance(configs);
+        verifyConfigsWorks(profilingService, tempDir);
+    }
+
+    @AfterEach
+    void tearDown() throws IOException {
+        profilingService.close();
+    }
+
+    @Test
+    public void testSingleInstance() throws IOException {
+        ProfilingService instance = ProfilingService.getInstance(configs);
+        Assertions.assertEquals(profilingService, instance);
+        instance.close();
+    }
+
+    @Test
+    void testFailedRequestUnderProfiling() throws ExecutionException, InterruptedException {
+        ProfilingInfo profilingInfo =
+                profilingService
+                        .requestProfiling(resourceID, 10, ProfilingInfo.ProfilingMode.ITIMER)
+                        .get();
+        Assertions.assertEquals(ProfilingInfo.ProfilingStatus.RUNNING, profilingInfo.getStatus());
+        try {
+            profilingService
+                    .requestProfiling(
+                            resourceID, profilingDuration, ProfilingInfo.ProfilingMode.ITIMER)
+                    .get();
+            Assertions.fail("Duplicate profiling request should throw with IllegalStateException.");
+        } catch (Exception e) {
+            Assertions.assertTrue(e.getCause() instanceof IllegalStateException);
+        }
+    }
+
+    @Test
+    @Timeout(value = 1, unit = TimeUnit.MINUTES)
+    public void testAllProfilingMode() throws ExecutionException, InterruptedException {
+        for (ProfilingInfo.ProfilingMode mode : ProfilingInfo.ProfilingMode.values()) {
+            ProfilingInfo profilingInfo =
+                    profilingService.requestProfiling(resourceID, profilingDuration, mode).get();
+            if (isNoPermissionOrAllocateSymbol(profilingInfo)) {
+                LOG.warn(
+                        "Ignoring failed profiling instance in {} mode, which caused by no permission.",
+                        profilingInfo.getProfilingMode());
+                continue;
+            }
+            Assertions.assertEquals(
+                    ProfilingInfo.ProfilingStatus.RUNNING,
+                    profilingInfo.getStatus(),
+                    String.format(
+                            "Submitting profiling request should be succeed or no permission, but got errorMsg=%s",
+                            profilingInfo.getMessage()));
+            waitForProfilingFinished(profilingService);
+            Assertions.assertEquals(
+                    ProfilingInfo.ProfilingStatus.FINISHED,
+                    profilingInfo.getStatus(),
+                    String.format(
+                            "Profiling request should complete successful, but got errorMsg=%s",
+                            profilingInfo.getMessage()));
+        }
+
+        verifyRollingDeletion(profilingService);
```

Comment:
It's better to introduce another unit test to only test the rolling deletion. 
Let one unit test just do one thing.

Reply:
Since rolling deletion is a follow-up action after sampling, it only happens when the number of Profiling Results exceeds the parameter limit. So I think it's more suitable to check it in the profiling test, so as to avoid re-sampling several times to construct the rolling deletion case specifically, WDYT?

Reply:
When we write the unit tests, we should obey the Single Responsibility Principle, which should just let one UT just focus one thing.

Reply:
Agree with you, I've checked the tests to make sure they follow a single goal.